xarray/backends/netCDF4_.py

```diff
@@ -16,7 +16,6 @@
     BackendEntrypoint,
     WritableCFDataStore,
     _normalize_path,
-    _open_datatree_netcdf,
```

Comment:
I think if you move these changes back to `backends/common` and leave them in `_open_datatree_netcdf`, You might be able to import the store for both the legacyhdf5 and the netcdf4 libraries. and then call `_open_datatree_netcdf` for both legacyhdf and netcdf.

currently `_open_datatree_netcdf`takes `ncDataset: ncDataset | ncDatasetLegacyH5,`

you might be able to include a new param with type `cdfDataStore: NetCDF4DataStore | H5NetCDFStore` and pass the appropritate one from both the n5netcdf_.py and netCDF4_.py backends.

Reply:
Hi @flamingbear, 

Following up on your comments, I would like to ask: Does `datatree` support the `to_h5` method? I was looking for this method implementation or `test` but didn't find it. Therefore, do we need to support the `open_datatree` method for `h5` files even if we do not have a `to_h5`?

Please let me know your thoughts.

Reply:
@aladinor I think @flamingbear will have a much more detailed answer, so here just some short thoughts in advance. Xarray is able to use `netCDF4` and `h5netcdf` to read (almost) any HDF5 and netCDF4 files (a netCDF4 file is essentially an HDF5 file). `netcdf4-python` is able to read non-conforming files ( as well as h5netcdf). Same when writing, `engine="netcdf4"` and `engine="h5netcdf"` will create similar (netCDF4) files. But here files will be standard conforming (beside using engine `h5netcdf` with `invalid_netcdf=True`).

Reply:
Yes, Kai is correct.  The h5netcdf is just a legacy/alternative interface to the netcdf4 files.  Both libraries supported a `Dataset` class with similar interfaces. The original datatree implementation [just chose which `Dataset` class](https://github.com/xarray-contrib/datatree/blob/0afaa6cc1d6800987d8b9c37a604dc0a8c68aeaa/datatree/io.py#L22-L34) to use based on which library was available and used them the same way.  Let me know if this helps or you want more info?

Reply:
Thanks, @kmuehlbauer and @flamingbear, for your feedback. I have one more question: Can I use the proposed implementation to open datatree stored in NetCDF4 files for H5 files?

Reply:
Hi, I was just looking at this now.  I was hoping that we could pass just the class `NetCDF4DataStore` or `H5NetCDFStore` to the common function and just `GenericStore.open()`, but it looks like there are different keywords to the open function.  
I was just checking to see if in NetCDF4_.open_datatree you can open the store, and pass that and the class to a common open_datatree.

``` python
        filename_or_obj = _normalize_path(filename_or_obj)
        store = NetCDF4DataStore.open(...)
        return _open_datatree_netcdf_common(NetCDF4DataStore, store, ...rest)
```

And in h5netcdf_.open_datatree:

``` python

        filename_or_obj = _normalize_path(filename_or_obj)
        store = H5NetCDFStore.open(...)
        return _open_datatree_netcdf_common(H5NetCDFStore, store, ...rest)
```

though, I haven't got it completed yet or determined if the different open() functions could be aligned with the same keywords.

Reply:
Thanks, @flamingbear, for taking a look at it. I also feel more confident about keeping them separately, as we must create a `group_store` when looping through each node.

```python 
for group in _iter_nc_groups(store.ds):
            group_path = str(parent / group[1:])
            group_store = NetCDF4DataStore(manager, group=group_path, **kwargs) #  <--- here
            store_entrypoint = StoreBackendEntrypoint()
```

 I created separate implementations to ensure all keywords are aligned with each `store` object.

Reply:
I'm trying not to harp on this too much, but that `group_store` you create has the same arguments whether it's a `NetCDF4DataStore` or a `H5NetCDFStore` which could be passed as a generic store.  I'm still on the fence if this is easier to maintain because it's in one place or if it's just adding complexity to try to reuse the code.

Reply:
I'll let someone else make this call. 

The only diffs I found in the Store.open() keywords are:

 - netcdf:
   `clobber`, `diskless`, `persist`, `autoclose`

- h5netcdf:
   `invalid_netcdf`, `phony_dims`, `decode_vlen_strings`, `driver`, `driver_kwds`
   
   But I'm a little naive about the named keywords vs **kwargs, and how I'd reconcile those.